task/dump-agent/src/main.rs

```diff
@@ -20,83 +20,48 @@ use userlib::*;
 const_assert!(DUMP_READ_SIZE <= 1024);
 
-struct ServerImpl {
-    areas: [DumpArea; 3],
-}
+struct ServerImpl;
 
 #[cfg(not(feature = "no-rot"))]
 task_slot!(SPROT, sprot);
 
+task_slot!(JEFE, jefe);
+
 impl ServerImpl {
-    fn initialize(&self) {
-        let mut next = 0;
-
-        for area in self.areas.iter().rev() {
-            unsafe {
-                let header = area.address as *mut DumpAreaHeader;
-
-                //
-                // We initialize our dump header with deliberately bad magic
-                // to prevent any dumps until we have everything initialized
-                //
-                (*header) = DumpAreaHeader {
-                    magic: DUMP_UNINITIALIZED,
-                    address: area.address,
-                    nsegments: 0,
-                    written: size_of::<DumpAreaHeader>() as u32,
-                    length: area.length,
-                    agent_version: DUMP_AGENT_VERSION,
-                    dumper_version: DUMPER_NONE,
-                    next,
-                }
-            }
+    fn initialize(&self) -> Result<(), DumpAgentError> {
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.initialize_dump_areas()
+    }
 
-            next = area.address;
-        }
+    fn dump_area(&self, index: u8) -> Result<DumpArea, DumpAgentError> {
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.get_dump_area(index)
+    }
 
-        for area in &self.areas {
-            unsafe {
-                let header = area.address as *mut DumpAreaHeader;
-                (*header).magic = DUMP_MAGIC;
-            }
-        }
+    fn claim_dump_area(&self) -> Result<DumpArea, DumpAgentError> {
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.claim_dump_area()
     }
 
     fn add_dump_segment(
         &mut self,
         addr: u32,
         length: u32,
     ) -> Result<(), DumpAgentError> {
-        let area = self.areas[0];
-
-        unsafe {
-            let header = area.address as *mut DumpAreaHeader;
-
-            if (*header).magic != DUMP_MAGIC {
-                panic!("bad dump magic!");
-            }
-
-            let nsegments = (*header).nsegments;
+        let area = self.dump_area(0)?;
 
-            let offset = size_of::<DumpAreaHeader>()
-                + (nsegments as usize) * size_of::<DumpSegmentHeader>();
-            let need = (offset + size_of::<DumpSegmentHeader>()) as u32;
-
-            if need > (*header).length {
-                return Err(DumpAgentError::OutOfSpaceForSegments);
-            }
-
-            let saddr = area.address as usize + offset;
-            let segment = saddr as *mut DumpSegmentHeader;
-
-            (*segment).address = addr;
-            (*segment).length = length;
-
-            (*header).nsegments = nsegments + 1;
-            (*header).written = need;
+        if area.agent != humpty::DumpAgent::Task {
+            self.claim_dump_area()?;
```

Comment:
Looking at the `humpty` code, it seems like claiming a dump area only modifies the `agent` field, but doesn't clear anything else.  If dump area 0 was previously used as a `Jefe` dump site, then it will have non-zero `nsegments` / `written`, which seems bad.

Reply:
Added a comment to clarify what's going on here.